test/move_effect_fling.c

```diff
@@ -0,0 +1,336 @@
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_PROTECT); MOVE(player, MOVE_FLING);}
+        TURN { MOVE(player, MOVE_FLING); }
+    } SCENE {
+        MESSAGE("Foe Wobbuffet used Protect!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_PROTECT, opponent);
+        MESSAGE("Wobbuffet used Fling!");
+        MESSAGE("Foe Wobbuffet protected itself!");
+
+        MESSAGE("Wobbuffet used Fling!");
+        MESSAGE("But it failed!");
+    } THEN {
+        EXPECT_EQ(player->item, ITEM_NONE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Fling doesn't consume the item if pokemon is asleep/frozen/paralyzed")
+{
+    u32 status;
+    u16 item;
+
+    PARAMETRIZE {status = STATUS1_SLEEP_TURN(2); item = ITEM_RAZOR_CLAW; }
+    PARAMETRIZE {status = STATUS1_PARALYSIS; item = ITEM_RAZOR_CLAW; }
+    PARAMETRIZE {status = STATUS1_FREEZE; item = ITEM_RAZOR_CLAW; }
+    PARAMETRIZE {status = STATUS1_SLEEP_TURN(2); item = ITEM_NONE; }
+    PARAMETRIZE {status = STATUS1_PARALYSIS; item = ITEM_NONE; }
+    PARAMETRIZE {status = STATUS1_FREEZE; item = ITEM_NONE; }
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) {Item(item); Status1(status); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        if (status == STATUS1_FREEZE) {
+            TURN { MOVE(player, MOVE_FLING, WITH_RNG(RNG_FROZEN, FALSE)); }
+            TURN { MOVE(player, MOVE_FLING, WITH_RNG(RNG_FROZEN, TRUE)); }
+        } else if (status == STATUS1_PARALYSIS) {
+            TURN { MOVE(player, MOVE_FLING, WITH_RNG(RNG_PARALYSIS, FALSE)); }
+            TURN { MOVE(player, MOVE_FLING, WITH_RNG(RNG_PARALYSIS, TRUE)); }
+        } else {
+            TURN { MOVE(player, MOVE_FLING); }
+            TURN { MOVE(player, MOVE_FLING); }
+        }
+    } SCENE {
+        if (status == STATUS1_FREEZE) {
+            MESSAGE("Wobbuffet is frozen solid!");
+            MESSAGE("Wobbuffet was defrosted!");
+        }
+        else if (status == STATUS1_PARALYSIS) {
+            MESSAGE("Wobbuffet is paralyzed! It can't move!");
+        }
+        else {
+            MESSAGE("Wobbuffet is fast asleep.");
+            MESSAGE("Wobbuffet woke up!");
+        }
+        MESSAGE("Wobbuffet used Fling!");
+        if (item != ITEM_NONE) {
+            ANIMATION(ANIM_TYPE_MOVE, MOVE_FLING, player);
+            HP_BAR(opponent);
+        } else {
+            MESSAGE("But it failed!");
+        }
+
+    } THEN {
+        EXPECT_EQ(player->item, ITEM_NONE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Fling applies special effects when throwing specific Items")
+{
+    u16 item, effect;
+
+    PARAMETRIZE {item = ITEM_FLAME_ORB; effect = EFFECT_WILL_O_WISP; }
+    PARAMETRIZE {item = ITEM_TOXIC_ORB; effect = EFFECT_TOXIC; }
+    PARAMETRIZE {item = ITEM_POISON_BARB; effect = EFFECT_POISON; }
+    PARAMETRIZE {item = ITEM_LIGHT_BALL; effect = EFFECT_PARALYZE; }
+    PARAMETRIZE {item = ITEM_RAZOR_FANG; effect = EFFECT_FLINCH_HIT; }
+    PARAMETRIZE {item = ITEM_KINGS_ROCK; effect = EFFECT_FLINCH_HIT; }
```

Comment:
Doesn't matter much, but using the `MOVE_EFFECT_x` might make more sense here
```suggestion
    PARAMETRIZE { item = ITEM_FLAME_ORB; effect = EFFECT_WILL_O_WISP; }
    PARAMETRIZE { item = ITEM_TOXIC_ORB; effect = EFFECT_TOXIC; }
    PARAMETRIZE { item = ITEM_POISON_BARB; effect = EFFECT_POISON; }
    PARAMETRIZE { item = ITEM_LIGHT_BALL; effect = EFFECT_PARALYZE; }
    PARAMETRIZE { item = ITEM_RAZOR_FANG; effect = EFFECT_FLINCH_HIT; }
    PARAMETRIZE { item = ITEM_KINGS_ROCK; effect = EFFECT_FLINCH_HIT; }
```